core/server/src/main/java/alluxio/worker/file/UnderFileSystemManager.java

```diff
@@ -153,33 +160,64 @@ private InputStreamAgent(long sessionId, long agentId, AlluxioURI ufsUri, Config
     }
 
     /**
-     * Opens a new input stream to the file this agent references. The new stream will be at the
-     * specified position when it is returned to the caller. The caller is responsible for
-     * closing the stream.
+     * Closes the internal stream of the input stream agent.
+     *
+     * @throws IOException if an error occurs when interacting with the UFS
+     */
+    private void close() throws IOException {
+      if (mStream != null) {
+        mStream.close();
+        mStream = null;
+      }
+    }
+
+    /**
+     * Checks if the current stream can be reused to serve the request. If so, the current
+     * stream is returned. Otherwise, opens a new input stream to the file this agent references
+     * and closes the previous stream. The new stream will be at the specified position when it
+     * is returned to the caller.
      *
      * @param position the absolute position in the file to start the stream at
      * @return an input stream to the file starting at the specified position, null if the position
      *         is past the end of the file
      * @throws IOException if an error occurs when interacting with the UFS
      */
     private InputStream openAtPosition(long position) throws IOException {
-      if (position >= mLength) {
+      if (position >= mLength) { // Position is at EOF
         return null;
       }
-      UnderFileSystem ufs = UnderFileSystem.get(mUri, mConfiguration);
-      // Special handling for S3/GCS with open object at position to save redundant ufs stream
-      // close and reopen in S3InputStream/GCSInputStream skip.
-      if (ufs instanceof S3UnderFileSystem) {
-        return ((S3UnderFileSystem) ufs).openAtPosition(mUri, position);
-      }
-      if (ufs instanceof GCSUnderFileSystem) {
-        return ((GCSUnderFileSystem) ufs).openAtPosition(mUri, position);
+
+      // If no stream has been created or if we need to go backward, make a new cached stream.
+      if (mStream == null || mInitPos + mStream.getCount() > position) {
+        if (mStream != null) { // Close the existing stream if needed
+          mStream.close();
+        }
+        UnderFileSystem ufs = UnderFileSystem.get(mUri, mConfiguration);
+        // TODO(calvin): Consider making openAtPosition part of the UFS API
```

Comment:
I think the change would be pretty minor, but I'm hesitant to change the ufs API in a patch version. Also I'm not sure if this is the correct API to provide. Although it solves this particular problem well, there may be a more general way of doing this which would solve a larger variety of problems. If the upside is mainly code maintenance, I think we can address this before 1.2, so this would only be in master for a short period of time.